File: site/system.py
"""HTML5 Ajax System function module"""
__author__= "Zacharias El Banna"

#
#
def main(aWeb):
 cookie = aWeb.cookie('rims')
 data = aWeb.rest_call("master/inventory",{'node':aWeb.node(),'user_id':cookie['id']})
 svcs = aWeb.rest_call("system/service_list")['services']
 tools = data.get('tools',[])
 aWeb.wr("<NAV><UL>")
 if aWeb.node() == 'master':
  aWeb.wr("<LI><A CLASS=z-op DIV=div_content URL='node_list'>Nodes</A></LI>")
  aWeb.wr("<LI><A CLASS=z-op DIV=div_content URL='server_list'>Servers</A></LI>")
  aWeb.wr("<LI><A CLASS=z-op TARGET=_blank   HREF='infra/erd.pdf'>ERD</A></LI>")
  aWeb.wr("<LI><A CLASS=z-op DIV=div_content URL='user_list'>Users</A></LI>")
 if data.get('logs'):
  aWeb.wr("<LI CLASS='dropdown'><A>Logs</A><DIV CLASS='dropdown-content'>")
  for node in data['logs']:
   aWeb.wr("<A CLASS=z-op DIV=div_content URL=system_logs_show?node=%s>%s</A>"%(node,node))
  aWeb.wr("</DIV></LI>")
 aWeb.wr("<LI CLASS='dropdown'><A>Reports</A><DIV CLASS='dropdown-content'>")
 if aWeb.node() == 'master':
  aWeb.wr("<A CLASS=z-op DIV=div_content URL='activity_report'>Activities</A>")
  aWeb.wr("<A CLASS=z-op DIV=div_content URL='reservation_report'>Reservations</A>")
  aWeb.wr("<A CLASS=z-op DIV=div_content URL='device_report'>Devices</A>")
  aWeb.wr("<A CLASS=z-op DIV=div_content URL='inventory_report'>Inventory</A>")
 aWeb.wr("<A CLASS=z-op DIV=div_content URL='system_task_report'>Tasks</A>")
 aWeb.wr("<A CLASS=z-op DIV=div_content URL='system_report'>System</A>")
 aWeb.wr("</DIV></LI>")
 aWeb.wr("<LI><A CLASS=z-op DIV=div_content URL='system_rest_explore'>REST</A></LI>")
 if len(tools) > 0 or len(svcs) > 0:
  aWeb.wr("<LI CLASS='dropdown'><A>Tools</A><DIV CLASS='dropdown-content'>")
  for tool in tools:
   aWeb.wr("<A CLASS=z-op DIV=div_content URL='%s'>%s</A>"%(tool['href'],tool['title']))
  for svc in svcs:
   aWeb.wr("<A CLASS=z-op DIV=div_content URL='system_services_info?service=%s'>%s</A>"%(svc['service'],svc['name']))
 aWeb.wr("</DIV></LI>")
 if aWeb.node() == 'master':
  aWeb.wr("<LI><A CLASS='z-op' DIV=div_content URL='system_controls'>Controls</A></LI>")
 aWeb.wr("<LI><A CLASS='z-op reload' DIV=main URL='system_main'></A></LI>")
 if data.get('navinfo'):
  for info in data['navinfo']:
   aWeb.wr("<LI CLASS='right navinfo'><A>%s</A></LI>"%info)
 aWeb.wr("</UL></NAV>")
 aWeb.wr("<SECTION CLASS=content ID=div_content></SECTION>")

#
#
def report(aWeb):
 info = aWeb.rest_full("%s/system/report"%aWeb.url(), aDataOnly = True)
 analytics = info.pop('analytics',{})
 keys = sorted(info.keys())
 aWeb.wr("<ARTICLE><P>System</P>")
 aWeb.wr("<DIV CLASS=table><DIV CLASS=thead><DIV>Information</DIV><DIV>Value</DIV></DIV><DIV CLASS=tbody>")
 for i in keys:
  aWeb.wr("<DIV><DIV>%s</DIV><DIV>%s</DIV></DIV>"%(i,info[i]))
 for x in ['modules','files']:
  for k,v in analytics.get(x,{}).items():
   aWeb.wr("<DIV><DIV>%s: %s</DIV><DIV>%s</DIV></DIV>"%(x.title(),k,v))
 aWeb.wr("</DIV></DIV></ARTICLE>")

#
#
def reload(aWeb):
 """ Map node to URL and call reload """
 api = aWeb.rest_call("system/node_to_api",{'node':aWeb['node']})['url']
 res = aWeb.rest_full("%s/system/reload"%api, aDataOnly = True)
 aWeb.wr("<ARTICLE><P>Module</P>")
 aWeb.wr("<DIV CLASS=table><DIV CLASS=tbody>")
 for x in res['modules']:
  aWeb.wr("<DIV><DIV>%s</DIV></DIV>"%x)
 aWeb.wr("</DIV></DIV></ARTICLE>")

######################################### Tasks ######################################
#
#
def task_report(aWeb):
 res = aWeb.rest_call("master/task_list",{'node':aWeb.node()})
 aWeb.wr("<ARTICLE><P>Tasks</P>")
 aWeb.wr("<DIV CLASS=table><DIV CLASS=thead><DIV>Node</DIV><DIV>Frequency</DIV><DIV>Module</DIV><DIV>Function</DIV><DIV>Args</DIV></DIV><DIV CLASS=tbody>")
 for task in res['tasks']:
  aWeb.wr("<DIV><DIV>%(node)s</DIV><DIV>%(frequency)s</DIV><DIV>%(module)s</DIV><DIV>%(function)s</DIV><DIV>%(args)s</DIV></DIV>"%task)
 aWeb.wr("</DIV></DIV></ARTICLE>")

############################################ Options ##############################################
#
#
#
def rest_explore(aWeb):
 res = aWeb.rest_call("system/rest_explore")
 aWeb.wr("<SECTION CLASS=content-left  ID=div_content_left>")
 aWeb.wr("<ARTICLE>")
 aWeb.wr("<DIV CLASS=table><DIV CLASS=thead><DIV>API</DIV><DIV>Function</DIV></DIV><DIV CLASS=tbody>")
 for item in res['data']:
  for fun in item['functions']:
   aWeb.wr("<DIV><DIV>%s</DIV><DIV><A CLASS=z-op DIV=div_content_right URL=system_rest_information?api=%s&function=%s>%s</A></DIV></DIV>"%(item['api'],item['api'],fun,fun))
 aWeb.wr("</DIV></DIV></ARTICLE>")
 aWeb.wr("</SECTION>")
 aWeb.wr("<SECTION CLASS=content-right ID=div_content_right></SECTION>")

#
#
def rest_information(aWeb):
 res = aWeb.rest_call("system/rest_information",{'api':aWeb['api'],'function':aWeb['function']})
 aWeb.wr("<ARTICLE>")
 aWeb.wr("<H1>API: %s</H1>"%(aWeb['api']))
 aWeb.wr("<BR>".join(res['module']))
 aWeb.wr("<H1>Function: %s</H1>"%(aWeb['function']))
 aWeb.wr("<BR>".join(res['information']))
 aWeb.wr("</ARTICLE>")

#
#
def rest_execute(aWeb):
 from json import dumps, loads
 args = loads(aWeb['arguments']) if 'arguments' in aWeb.args() else None
 data = aWeb.rest_full("%s/api/%s"%(aWeb.url(),aWeb['api']), aArgs = args, aTimeOut=300, aDataOnly = True)
 aWeb.wr("<ARTICLE><PRE CLASS='white'>%s</PRE></ARTICLE>"%dumps(data,indent=4, sort_keys=True))


############################################### Logs ###############################################

#
#
def logs_clear(aWeb):
 args = aWeb.args()
 node = args.pop('node',None)
 args['count'] = 18
 res = aWeb.rest_call('system/logs_clear%s'%('?node=%s'%node if node else ''),args)
 aWeb.wr("<ARTICLE><P>%s</P>"%res['node'])
 for i in res['file'].items():
  aWeb.wr("%s: %s<BR>"%i)
 aWeb.wr("</ARTICLE>"%(res))

#
#
def logs_show(aWeb):
 args = aWeb.args()
 node = args.pop('node',None)
 args['count'] = 18
 res = aWeb.rest_call('system/logs_get%s'%('?node=%s'%node if node else ''),args)
 aWeb.wr("<ARTICLE>")
 for file,logs in res.items():
  aWeb.wr("<P STYLE='font-weight:bold; text-align:center;'>%s</P><P CLASS='machine-text'>%s</P>"%(file,"<BR>".join(logs)))
 aWeb.wr("</ARTICLE>")

############################################# Services ##############################################
#
#
def services_info(aWeb):
 args = aWeb.args()
 data  = aWeb.rest_call('system/service_info',args)
 state = 'start' if data['state'] == 'inactive' else 'stop'
 aWeb.wr("<ARTICLE STYLE='display:inline-block;'><B>%s</B>: %s (%s)"%(aWeb['service'],data['state'],data['extra']))
 aWeb.wr(aWeb.button(state, DIV='div_content', SPIN='true', URL='system_services_info?service=%s&op=%s'%(args['service'],state)))
 aWeb.wr("</ARTICLE>")

############################################## Files ###############################################
#
#
def file_list(aWeb):
 res = aWeb.rest_call('system/file_list',{'directory':aWeb['directory']})
 if aWeb.get('navigation'):
  aWeb.wr("<NAV><UL>&nbsp;</UL></NAV>")
  aWeb.wr("<SECTION CLASS=content ID=div_content>")
 aWeb.wr("<ARTICLE><P>Files in %s<P>"%res.get('path','directory'))
 import urllib.request, urllib.parse, urllib.error
 for f in res['files']:
  url  = urllib.parse.quote(f)
  aWeb.wr("<P CLASS=machine-text>{0}/<A HREF='{0}/{1}' TARGET=_blank>{2}</A></P>".format(res.get('path','#'),url,f))
 aWeb.wr("</ARTICLE>")
 if aWeb.get('navigation'):
  aWeb.wr("</SECTION>")

#
#
def images(aWeb):
 res = aWeb.rest_call('system/file_list')
 if aWeb.get('navigation'):
  aWeb.wr("<NAV><UL>&nbsp;</UL></NAV>")
  aWeb.wr("<SECTION CLASS=content ID=div_content>")
 aWeb.wr("<ARTICLE><P>Images<P><DIV CLASS=table><DIV CLASS=tbody>")
 for file in res['files']:
  if file[-3:] == 'png':
   aWeb.wr("<DIV><DIV STYLE='max-width:180px'>{0}</DIV><DIV STYLE='width:100%;'><IMG STYLE='max-height:60px;' SRC='{1}/{0}' /></DIV></DIV>".format(file,res['path']))
 aWeb.wr("</DIV></DIV></ARTICLE>")
 if aWeb.get('navigation'):
  aWeb.wr("</SECTION>")

############################################ Controls ##############################################
#
#
def controls(aWeb):
 aWeb.wr("<SECTION CLASS=content-left ID=div_content_left>")
 aWeb.wr("<ARTICLE><DIV CLASS=table><DIV CLASS=tbody>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=monitor/ipam_init'>IPAM status check</A></DIV></DIV>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=monitor/ipam_events&arguments={\"op\":\"clear\"}'>IPAM clear status logs</A></DIV></DIV>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=monitor/interface_init'>Interface status check</A></DIV></DIV>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=device/network_info_discover'>Discover system info</A></DIV></DIV>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=device/model_sync'>Sync models</A></DIV></DIV>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=device/vm_mapping'>VM UUID mapping</A></DIV></DIV>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=reservation/expiration_status'>Reservation checks</A></DIV></DIV>")
 aWeb.wr("<DIV><DIV><A CLASS=z-op DIV=div_content_right SPIN='true' URL='system_rest_execute?api=master/oui_fetch'>OUI Database sync</A></DIV></DIV>")
 aWeb.wr("</DIV></DIV></ARTICLE>")
 aWeb.wr("</SECTION>")
 aWeb.wr("<SECTION CLASS=content-right ID=div_content_right></SECTION>")

File: src/System.js
import React, { Component } from 'react'
import { rest_call, rest_base, read_cookie } from './infra/Functions.js';
import { Spinner, ContentList } from './infra/Generic.js';
import { MainBase, ListBase, ReportBase } from './infra/Base.js';
import { TextButton } from './infra/Buttons.js';

import { LogShow, List as NodeList } from './Node.js';
import { List as ServerList } from './Server.js';
import { List as UserList } from './User.js';
import { Report as ActivityReport } from './Activity.js';
import { Report as ReservationReport } from './Reservation.js';
import { Report as DeviceReport } from './Device.js';
import { Report as InventoryReport } from './Inventory.js';

// ************** Main **************
//
export class Main extends MainBase {
 constructor(props){
  super(props)
  this.state.cookie = read_cookie('rims')
  this.state.navinfo = []
  this.state.navitems = []
  this.state.logs = []
  this.state.services = []
 }

 componentDidMount(){
  rest_call(rest_base + 'api/system/service_list')
   .then((result) => {
    Object.assign(this.state,result)
    this.compileNavItems()
   })
  rest_call(rest_base + 'api/master/inventory',{node:this.state.cookie.node, user_id:this.state.cookie.id})
   .then((result) => {
    Object.assign(this.state,result)
    this.compileNavItems()
   })
 }

 compileNavItems = () => {
  let navitems = [];
  let reports = [];
   if (this.state.cookie.node === 'master') {
    navitems.push({title:'Nodes',  onClick:() => this.changeMain(<NodeList key='node_list' />) })
    navitems.push({title:'Servers', onClick:() => this.changeMain(<ServerList key='server_list' />) })
    navitems.push({title:'ERD',    onClick:() => window.open('infra/erd.pdf','_blank') })
    navitems.push({title:'Users',  onClick:() => this.changeMain(<UserList key='user_list' />) })
    navitems.push({title:'Controls',  onClick:() => this.changeMain(<Controls  />) })
    reports.push({title:'Activities',  onClick:() => this.changeMain(<ActivityReport key='activity_report' />) })
    reports.push({title:'Reservations',  onClick:() => this.changeMain(<ReservationReport key='reservation_report' />) })
    reports.push({title:'Devices',  onClick:() => this.changeMain(<DeviceReport key='device_report' />) })
    reports.push({title:'Inventory',  onClick:() => this.changeMain(<InventoryReport key='inventory_report' />) })
   }
   reports.push({title:'Tasks',   onClick:() => this.changeMain(<TaskReport key='task_report' />) })
   reports.push({title:'System',  onClick:() => this.changeMain(<Report key='system_report' />) })
   navitems.push({title:'Logs',   type:'dropdown', items:this.state.logs.map((row,index) => { return {title:row, onClick:() => this.changeMain( <LogShow node={row} /> ) }} ) })
   navitems.push({title:'Report', type:'dropdown', items:reports})
   navitems.push({title:'REST',  onClick:() => this.changeMain(<RestList key='rest_list' />) })
   if (this.state.services.length > 0)
    navitems.push({title:'Services', type:'dropdown', items:this.state.services.map((row,index) => { return {title:row, onClick:() => this.changeMain( <ServiceInfo {... row} /> ) }} ) })
   navitems.push({ onClick:() => this.setState({content:null}), className:'reload' })
   this.state.navinfo.forEach(row => navitems.push({title:row, className:'right navinfo'}))
   this.props.loadNavigation(navitems)
 }

}

// ************** Report **************
//
export class Report extends ReportBase{
 constructor(props){
  super(props)
  this.header = 'System Report'
  this.thead = ['Key','Value']
 }

 componentDidMount(){
  rest_call(rest_base + 'system/report')
   .then((result) => {
   // TODO: Does not work...
   this.setState({data:Object.keys(result).sort((a,b) => { return (b - a) }).map(key => { return {info:key,value:result[key]}})});
  })
 }

 listItem = (row) => [row.info,row.value]

}

// ************** TaskReport **************
//
export class TaskReport extends ReportBase {
 constructor(props){
  super(props);
  const cookie = read_cookie('rims')
  this.state.node = cookie.node
  this.header = 'Tasks'
  this.thead = ['Node','Frequency','Module','Function','Arguments']
 }

 componentDidMount(){
  rest_call(rest_base + 'api/master/task_list',{node:this.state.node})
   .then((result) => { this.setState(result) })
 }

 listItem = (row) => [row.node,row.frequency,row.module,row.function,row.args]

}

// ************** RestList **************
//
class RestList extends ListBase {
 constructor(props){
  super(props);
  this.thead=['API','Function']
  this.header='REST API'
 }

 componentDidMount(){
  rest_call(rest_base + 'api/system/rest_explore')
   .then((result) => {
    var apilist = [];
    result.data.forEach(item => item.functions.forEach(row => apilist.push({api:item.api, function:row})  ) );
    this.setState({data:apilist});
   })
 }

 listItem = (row) => [row.api,<TextButton key={'rest_' + row.api} text={row.function} onClick={() => { this.changeList(<RestInfo key={`rest_info_${row.api}_${row.function}`} {...row} />)}} />]

}


// ************** RestInfo **************
//
class RestInfo extends Component {
 constructor(props){
  super(props)
  this.state = {module:null,information:null}
 }

 componentDidMount(){
  rest_call(rest_base + 'api/system/rest_information',{api:this.props.api, function:this.props.function})
   .then((result) => { this.setState(result) })
 }

 render() {
  if ((this.state.module === null) || (this.state.information === null))
   return <Spinner />
  else {
   return(
    <article className='text'>
     <h1>API: {this.props.api}</h1>
     {this.state.module.map((row,index) => {return <p key={'ria_'+index}>{row}</p> })}
     <h1>Function: {this.props.function}</h1>
     {this.state.information.map((row,index) => {return <p key={'rif_'+index}>{row}</p> })}
    </article>
   )
  }
 }
}

// ************************ Controls ********************
//
class RestExecute extends Component {
 componentDidMount(){
  const args = ('args' in this.props) ? this.props.args : {}
  rest_call(rest_base + 'api/' + this.props.api,args)
  .then((result) => { this.setState(result) })
 }

 render(){
  if (this.state === null)
   return <Spinner />
  else
   return <article className='code'><h1>{this.props.text}</h1><pre>{JSON.stringify(this.state,null,2)}</pre></article>
 }
}

// ************************ Controls ********************
//
// TODO: List should be dynamic from config and passed through REST engine
class Controls extends ListBase {
 constructor(props){
  super(props);
  this.state.data = [
   {api:'monitor/ipam_init',text:'IPAM status check'},
   {api:'ipam/address_events',text:'IPAM clear status logs',args:{op:'clear'}},
   {api:'monitor/interface_init',text:'Interface status check'},
   {api:'device/network_info_discover',text:'Discover device system information (sysmac etc)'},
   {api:'device/model_sync',text:'Sync device model mapping'},
   {api:'device/vm_mapping',text:'VM UUID mapping'},
   {api:'master/oui_fetch',text:'Sync OUI database'},
   {api:'reservation/expiration_status',text:'Check reservation status'}
  ]
  this.header = undefined
  this.thead = ['Control Function']
 }

 listItem = (row) => [<TextButton key={'ctrl_' + row.api} text={row.text} onClick={() => { this.changeList(<RestExecute key={'rest_' + row.api} {...row} />)}} />]
}

// ************************ TODO ********************

class ServiceInfo extends Component {
 render() {
  return (<div>Service Info TODO</div>)
 }
}

export class FileList extends Component {
 render() {
  return(<div>FileList</div>)
 }
}

export class Images extends Component {
 render() {
  return(<div>Images</div>)
 }
}
